### PR TITLE
Fix fire in window view with hop window

### PR DESCRIPTION
**Pull request by Vxider**
### Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Fix fire in window view with hop window #34044

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Storages/WindowView/StorageWindowView.cpp [hunk 1]
```diff
@@ -909,38 +909,36 @@ void StorageWindowView::threadFuncCleanup()
...
             tryLogCurrentException(__PRETTY_FUNCTION__);
         }
-        max_fired_watermark = exact_fire_signal;
-        next_fire_signal = addTime(next_fire_signal, window_kind, window_num_units, *time_zone);
-        exact_fire_signal = window_kind_larger_than_day ? next_fire_signal * 86400 : next_fire_signal;
+        max_fired_watermark = next_fire_signal;
+        auto slide_interval
+            = is_tumble ? addTime(0, window_kind, window_num_units, *time_zone) : addTime(0, hop_kind, hop_num_units, *time_zone);
+        /// When window is larger than day, getWindowUpperBound() will get a day num instead of timestamp,
+        /// and addTime will also add day num to the next_fire_signal, so we need to convert it into timestamp.
```
Comment (serxa) [inline src/Storages/WindowView/StorageWindowView.cpp:929]: now this comment is wrong, because next_fire_signal always contains seconds

Comment (serxa) [inline src/Storages/WindowView/StorageWindowView.cpp:929]: now this comment is wrong, because next_fire_signal always contains seconds

**Review by serxa — CHANGES_REQUESTED** [review]

**Comment by serxa** [comment]
Btw what is the reason for using days instead of seconds starting from week in the first place?

File: src/Storages/WindowView/StorageWindowView.cpp (hunk 1 above)
Comment (Vxider) [reply src/Storages/WindowView/StorageWindowView.cpp]: Comments are modified.

**Comment by Vxider** [comment]
> Btw what is the reason for using days instead of seconds starting from week in the first place?

Because `toStartOfXXXInterval` returns `DayNum` when interval is larger than Day,  we can also convert DayNum into seconds in `ToStartOfTransform<IntervalKind::INTERVAL_KIND>`(in `FunctionsTimeWindow.h`) but this will cost more storage space.

File: src/Storages/WindowView/StorageWindowView.cpp (hunk 1 above)
Comment (Vxider) [reply src/Storages/WindowView/StorageWindowView.cpp]: Comments are modified.

File: src/Storages/WindowView/StorageWindowView.cpp [hunk 2]
```diff
@@ -1213,11 +1181,8 @@ void StorageWindowView::writeIntoWindowView(
 
         if (window_view.is_watermark_bounded)
         {
-            UInt32 watermark_lower_bound = window_view.is_tumble
-                ? addTime(t_max_watermark, window_view.window_kind,
-                          -1 * window_view.window_num_units, *window_view.time_zone)
-                : addTime(t_max_watermark, window_view.hop_kind,
-                          -1 * window_view.hop_num_units, *window_view.time_zone);
+            UInt32 watermark_lower_bound
+                = addTime(t_max_watermark, window_view.slide_kind, -1 * window_view.slide_num_units, *window_view.time_zone);
```
Comment (serxa) [inline src/Storages/WindowView/StorageWindowView.cpp:1185]: better?

```suggestion
                = addTime(t_max_watermark, window_view.slide_kind, -window_view.slide_num_units, *window_view.time_zone);
```

**Comment by serxa** [comment]
> > Btw what is the reason for using days instead of seconds starting from week in the first place?
> 
> Because `toStartOfXXXInterval` returns `DayNum` when interval is larger than Day, we can also convert DayNum into seconds in `ToStartOfTransform<IntervalKind::INTERVAL_KIND>`(in `FunctionsTimeWindow.h`) but this will cost more storage space.

Okay, let's not change behavior.

File: src/Storages/WindowView/StorageWindowView.cpp (hunk 2 above)
Comment (Vxider) [reply src/Storages/WindowView/StorageWindowView.cpp]: Thanks, all similar code have been updated.

**Review by serxa — APPROVED** [review]

**Comment by serxa** [comment]
Looks like fails are unrelated, but lets check one more time

**Comment by serxa** [comment]
@Mergifyio update

**Pull request merged**